src/Hazelcast.Net/CP/FencedLock.cs

```diff
@@ -0,0 +1,394 @@
+
+            var ownership = await RequestLockOwnershipStateAsync().CfAwait();
+
+            if (ownership.LockedBy(threadId, sessionId))
+                _lockedThreadToSession[threadId] = sessionId;
+            else
+                VerifyNoLockedSessionExist(threadId);
+
+            return ownership.LockCount;
+        }
+
+        /// <inheritdoc/>  
+        public async Task<bool> IsLockedAsync()
+        {
+            var threadId = ContextId;
+            var sessionId = _subsystemSession.GetSessionId(CPGroupId);
+
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            var ownership = await RequestLockOwnershipStateAsync().CfAwait();
+
+            if (ownership.LockedBy(threadId, sessionId))
+            {
+                _lockedThreadToSession[threadId] = sessionId;
+                return true;
+            }
+
+            VerifyNoLockedSessionExist(threadId);
+
+            return ownership.Locked;
+        }
+
+        public async Task<bool> IsLockedByCurrentContext()
+        {
+            var threadId = ContextId;
+            var sessionId = _subsystemSession.GetSessionId(CPGroupId);
+
+            VerifyNoLockOnThread(threadId, sessionId, false);
+
+            var ownership = await RequestLockOwnershipStateAsync().CfAwait();
+
+            var lockedByCurrent = ownership.LockedBy(threadId, sessionId);
+
+            if (lockedByCurrent)
+                _lockedThreadToSession[threadId] = sessionId;
+            else
+                VerifyNoLockedSessionExist(threadId);
+
+            return lockedByCurrent;
+        }
+
+        /// <inheritdoc/>        
+        public async Task<long> LockAndGetFenceAsync()
+        {
+            var threadId = ContextId;
+            var invocationId = Guid.NewGuid();// required by server, to make the call idempotetent?
+
+            while (true)
+            {
+                var sessionId = await _subsystemSession.AcquireSessionAsync(CPGroupId).CfAwait();
+
+                VerifyNoLockOnThread(threadId, sessionId, true);
+
+                try
+                {
+                    long fence = await RequestLockAsync(sessionId, threadId, invocationId).CfAwait();
+
+                    if (fence != InvalidFence)
+                    {
+                        _lockedThreadToSession[threadId] = sessionId;
+                        return fence;
+                    }
+
+                    throw new LockAcquireLimitReachedException($"Lock[{Name}] reentrant lock limit is already reached!");
+
+                }
+                catch (RemoteException e)
+                {
+                    if (e is RemoteException { Error: RemoteError.SessionExpiredException })
+                    {
+                        _subsystemSession.InvalidateSession(CPGroupId, sessionId);
+                        VerifyNoLockedSessionExist(threadId);
+                    }
+                    else if (e is RemoteException { Error: RemoteError.DistributedObjectDestroyed })
```

Comment:
why do you do specific to `DistributedObjectDestroyed `? You need to do it for all the rest of the exceptions. And you also need to do `_subsystemSession.ReleaseSession(CPGroupId, sessionId);` before re-throwing.

Reply:
remove DistributedObjectDestroyed? waitkeycancelled case should throw IllegalMonitorStateException as in Java.

Reply:
As current behavior, throws only `RemoteException. `Fixed at https://github.com/hazelcast/hazelcast-csharp-client/pull/590/commits/993cabbcb9bac708cc38ea83239b98cf19ead875

Reply:
For WaitKeyCancelled , please change it to a new RemoteException with error code IllegalStateError similar to  Java. It changes the rethrown exception.